### PR TITLE
feat: support jstype custom options

### LINKED ISSUE #112: Support 64 bit int/fixed as number
In order to be performant in certain usecases, we need to be able to force generated code to use `number` instead of `bigint`.

Protons should be extended to support a field option named `jstype` used like so:

```proto
message Foo {
  uint64 bar = 1 [jstype = JS_NUMBER];
}
```
```ts
import { Foo } from './foo.js'

const foo = {
  bar: 100 // a number :)
}

typeof Foo.decode(Foo.encode(foo)).bar === 'number'
```

See 
- https://protobuf.dev/programming-guides/proto3/#customoptions
- https://protobuf.dev/programming-guides/proto2/#options
- https://github.com/protocolbuffers/protobuf/blob/f881c8fe15b8c9aa3be997103eee5f2bea02984d/src/google/protobuf/descriptor.proto#L620-L641
- https://github.com/ChainSafe/js-libp2p-gossipsub/pull/327#issuecomment-1679857655

### PR DESCRIPTION
**Pull request by achingbrain**
Allow overriding 64 bit number types with strings or regular js numbers.

Closes #112

### CODE REVIEW

File: packages/protons/src/index.ts [hunk 1]
```diff
@@ -55,6 +60,11 @@ const encoderGenerators: Record<string, (val: string) => string> = {
   uint64: (val) => `w.uint64(${val})`
 }
 
+const encoderGeneratorsJsTypeOverrides: Record<string, (val: string) => string> = {
+  number: (val: string) => `BigInt(${val})`,
```
Comment (wemeetagain) [inline packages/protons/src/index.ts:64]: I haven't looked at exactly whats going on here but this doesn't seem ideal.
One goal of this feature is to take advantage of speedups available without using BigInt, but if we're going thru BigInt then that isn't happening.

Somehow I thought this feature would require creating some additional types eg `w.number64` where we could have optimized encoder/decoders.

Comment (achingbrain) [reply packages/protons/src/index.ts]: Ok, I've ported the original `LongBits.fromNumber` method from `protobuf.js` and split the codepaths out for bigints, numbers and strings so we have single call sites for future optimisations.

**Comment by achingbrain** [comment]
I've added a read/write number benchmark which shows the performance gap between numbers and bigints:

```
-- read --
uint64 (BigInt) x 64,510,135 ops/sec ±1.85% (89 runs sampled)
uint64number x 72,184,161 ops/sec ±0.63% (94 runs sampled)
uint64string x 31,305,471 ops/sec ±0.69% (90 runs sampled)
Fastest is uint64number
-- write --
uint64 (BigInt) x 20,341,970 ops/sec ±0.19% (97 runs sampled)
uint64number x 90,736,928 ops/sec ±0.46% (95 runs sampled)
uint64string x 12,260,985 ops/sec ±0.18% (101 runs sampled)
Fastest is uint64number
```

Based on that I've added a simple optimisation that downcasts a BigInt to a Number in `longbits.ts` if it's below `Number.MAX_SAFE_INTEGER` - it doubles the speed of writing smaller BigInt values:

```
-- read --
uint64 (BigInt) x 67,055,859 ops/sec ±1.72% (87 runs sampled)
uint64number x 73,642,770 ops/sec ±0.58% (95 runs sampled)
uint64string x 31,726,482 ops/sec ±0.47% (92 runs sampled)
Fastest is uint64number
-- write --
uint64 (BigInt) x 42,694,998 ops/sec ±0.57% (98 runs sampled)
uint64number x 86,652,331 ops/sec ±0.99% (91 runs sampled)
uint64string x 18,808,458 ops/sec ±0.15% (97 runs sampled)
Fastest is uint64number
```

**Comment by achingbrain** [comment]
I'm going to merge this because it improves a very real performance bottleneck.

**Pull request merged**